src/main/java/io/cdap/plugin/successfactors/source/input/SuccessFactorsPartitionBuilder.java

```diff
@@ -16,88 +16,48 @@
-   * @param fetchRowCount        plugin property, number of rows to extract. 0 (zero) means the total record to fetch
-   *                             will be same as total number of available records.
    * @return list of {@code SuccessFactorsInputSplit}
    */
-  public List<SuccessFactorsInputSplit> buildSplits(long availableRecordCount, long fetchRowCount) {
+  public List<SuccessFactorsInputSplit> buildSplits(long availableRecordCount) {
 
     List<SuccessFactorsInputSplit> list = new ArrayList<>();
-
-    long recordReadStartIndex = 1;
-    long actualRecordToExtract = (fetchRowCount == 0 ? availableRecordCount : fetchRowCount);
-    long recordReadEndIndex = Math.min(actualRecordToExtract, availableRecordCount);
-    if (actualRecordToExtract > availableRecordCount) {
-      actualRecordToExtract = availableRecordCount;
-      if (actualRecordToExtract <= 0) {
-        String msg = "As per the provided configuration no records were found for extraction.";
-        throw new IllegalArgumentException(msg);
-      }
-    }
-
-    // setting up the optimal package size values
-    long packageSize = Math.min(actualRecordToExtract, MAX_ALLOWED_PACKAGE_SIZE);
-
-    int splitCount = DEFAULT_SPLIT_COUNT;
-
-    // in case total fetch record count is less than the DEFAULT_PACKAGE_SIZE then Split count will be defaulted to 1.
-    if (actualRecordToExtract <= MAX_ALLOWED_PACKAGE_SIZE) {
-      splitCount = 1;
-    }
-
-    long optimalLoadOnSplit =
-      (splitCount == 1 ? actualRecordToExtract : (actualRecordToExtract / splitCount));
-    long optimalPackageSize = Math.min(optimalLoadOnSplit, packageSize);
-
-    long leftoverLoadCount = actualRecordToExtract % splitCount;
-
-    LOG.info("Total number to available record: {}", availableRecordCount);
-    LOG.info("Total number of record to extract: {}", actualRecordToExtract);
-    LOG.info("Record extraction to begin at index: {}", recordReadStartIndex);
-    LOG.info("Record extraction to end at index: {}", recordReadEndIndex);
-    LOG.info("Calculated number of splits: {}", splitCount);
-    LOG.info("Optimal record count to extract on the each splits: {}", optimalLoadOnSplit);
-    LOG.info("Optimal package size in each splits: {}", optimalPackageSize);
-
-    long start = recordReadStartIndex;
-
-    for (int i = 0; i < splitCount; i++) {
-      long extra = i < leftoverLoadCount ? 1 : 0;
+    long start = 1;
+    // setting up the optimal split size and count values
+    long batchSize = Math.min(availableRecordCount, MAX_ALLOWED_BATCH_SIZE);
+    long optimalLoadOnSplit = Math.min(availableRecordCount, MAX_RECORDS_IN_SPLIT);
+    long optimalSplitCount = availableRecordCount / optimalLoadOnSplit +
+      (availableRecordCount % optimalLoadOnSplit != 0 ? 1 : 0);
+    long leftoverLoadCount = availableRecordCount % optimalSplitCount;
+
+    for (int split = 1; split <= optimalSplitCount; split++) {
+      long extra = split < leftoverLoadCount ? 1 : 0;
```

Comment:
We can remove this logic of adding an extra record to the split. 
If you have N splits, then the first N-1 splits should have MAX_RECORDS_IN_SPLIT records. The last split should have <= MAX_RECORDS_IN_SPLIT records. So the load on N-1 splits will be evenly balanced. The last split might have lower load on it depending on availableRecordCount